include/blmc_robots/real_finger.hpp

```diff
@@ -115,105 +118,187 @@ class RealFinger : public robot_interfaces::Finger {
-      int linearly_decrease_time_steps = 1000;
-      int zero_torque_time_steps = 500;
-
-      /// \todo: this is yet another hack. we release the tip
-      /// joint before the others such that the tip joint and the center joint
-      /// dont fight each other
-      count = 0;
-      Vector torques;
-      while (count < linearly_decrease_time_steps) {
-        torques = -get_max_torques();
-
-        torques[2] = ((linearly_decrease_time_steps - count + 0.0) /
-                      linearly_decrease_time_steps) *
-                     torques[2];
-
-        TimeIndex t = append_desired_action(torques);
-        wait_until_time_index(t);
-        count++;
-      }
-      count = 0;
-      while (count < zero_torque_time_steps) {
-        TimeIndex t = append_desired_action(torques);
-        wait_until_time_index(t);
-        count++;
-      }
-
-      angle_offsets = get_observation(current_time_index()).angle;
-      joint_modules_.set_zero_angles(angle_offsets);
+        /// \todo: this relies on the safety check in the motor right now,
+        /// which is maybe not the greatest idea. Without the velocity and
+        /// torque limitation in the motor this would be very unsafe
+
+        //! Max. number of steps when searching for the encoder indices.
+        const uint32_t MAX_STEPS_SEARCH_INDEX = 1000;
+        //! Min. number of steps when moving to the end stop.
+        const uint32_t MIN_STEPS_MOVE_TO_END_STOP = 1000;
+        //! Size of the window when computing average velocity.
+        const uint32_t SIZE_VELOCITY_WINDOW = 100;
+        //! Velocity limit at which the joints are considered to be stopped.
+        const double STOP_VELOCITY = 0.001;
+
+
+        // First move a bit in positive direction until encoder indices are
+        // found or timeout triggers.
+        int count = 0;
+        while (joint_modules_.get_measured_index_angles().hasNaN()
+                && count < MAX_STEPS_SEARCH_INDEX) {
+            Vector torques = torque_ratio * get_max_torques();
+            TimeIndex t = append_desired_action(torques);
+            wait_until_time_index(t);
+            count++;
+        }
+        // Note: It might be possible that one or more encoder indices are not
+        // found before the timeout triggers (e.g. when joint is already at the
+        // positive end stop).  Still continue, as in this case the index might
+        // be found while moving to the negative end stop in the following step.
+        // If index is still not found for some reason, this is caught by a
+        // check at the end.
+
+
+        // Move until velocity drops to almost zero (= joints hit the end stops)
+        // but at least for MIN_STEPS_MOVE_TO_END_STOP time steps.
+        // TODO: add timeout to this loop?
+        std::vector<Vector> running_velocities(SIZE_VELOCITY_WINDOW);
+        Vector summed_velocities = Vector::Zero();
+        count = 0;
+        while (count < MIN_STEPS_MOVE_TO_END_STOP
```

Comment:
i think there is a slight problem here if MIN_STEPS_MOVE_TO_END_STOP < SIZE_VELOCITY_WINDOW:

the condition 
summed_velocities.maxCoeff() / SIZE_VELOCITY_WINDOW > STOP_VELOCITY

only makes sense when count >= SIZE_VELOCITY_WINDOW, because otherwise the summed_velocities does not actually contain SIZE_VELOCITY_WINDOW summands, and we are hence underestimating the average.